@here/harp-mapview/lib/text/LayoutState.ts

```diff
@@ -0,0 +1,137 @@
+    private m_hAlign = DefaultTextStyle.DEFAULT_HORIZONTAL_ALIGNMENT;
+    private m_vAlign = DefaultTextStyle.DEFAULT_VERTICAL_ALIGNMENT;
+
+    /**
+     * Set layout based on theme style defined and optional text placement.
+     *
+     * @param layoutStyle The basic theme label style.
+     * @param placement The optional new anchor placement.
+     */
+    setPlacement(layoutStyle: TextLayoutStyle, placement?: AnchorPlacement) {
+        if (placement !== undefined) {
+            this.m_hAlign = placement.h;
+            this.m_vAlign = placement.v;
+            this.m_state =
+                placement.h === layoutStyle.horizontalAlignment &&
+                placement.v === layoutStyle.verticalAlignment
+                    ? LayoutUsed.ThemeBased
+                    : LayoutUsed.Alternative;
+        } else {
+            this.m_hAlign = layoutStyle.horizontalAlignment;
+            this.m_vAlign = layoutStyle.verticalAlignment;
+            this.m_state = LayoutUsed.ThemeBased;
+        }
+    }
+
+    /**
+     * Acquire current placement setup.
+     *
+     * Function returns alternative or base placement depending on layout state.
+     *
+     * @param layoutStyle The label layout style.
+     * @returns The current anchor placement.
+     */
+    getPlacement(layoutStyle: TextLayoutStyle): AnchorPlacement {
+        if (this.isAlternative()) {
+            return { h: this.m_hAlign, v: this.m_vAlign };
+        } else {
+            return { h: layoutStyle.horizontalAlignment, v: layoutStyle.verticalAlignment };
+        }
+    }
+
+    /**
+     * Reset existing `LayoutState` to appear like a fresh state.
+     */
+    reset() {
+        this.m_state = LayoutUsed.Undefined;
+    }
+
+    /**
+     * @returns `true` if element has not yet been placed and layout variant assigned.
+     */
+    isUndefined(): boolean {
+        return this.m_state === LayoutUsed.Undefined;
+    }
+
+    /**
+     * @returns `true` if element is using original style defined in theme.
+     */
+    isBase(): boolean {
+        return this.m_state === LayoutUsed.ThemeBased;
+    }
+
+    /**
+     * @returns `true` if element is fading in.
+     */
+    isAlternative(): boolean {
+        return this.m_state === LayoutUsed.Alternative;
+    }
+
+    get horizontalAlignment(): HorizontalAlignment {
+        return this.m_hAlign;
+    }
+
+    get verticalAlignment(): VerticalAlignment {
+        return this.m_vAlign;
+    }
+
+    /**
+     * Clone this `LayoutState`.
+     *
+     * @returns Cloned `LayoutState`.
+     */
+    clone(): LayoutState {
+        const res = new LayoutState();
+        res.copy(this);
+        return res;
+    }
+
+    /**
+     * Copies the values of `other` to this [[LayoutState]].
+     *
+     * @param other The other [[LayoutState]] to copy.
+     */
+    copy(other: LayoutState) {
+        this.m_hAlign = other.m_hAlign;
+        this.m_vAlign = other.m_vAlign;
+        this.m_state = other.m_state;
+    }
```

Comment:
Remove this code if not used.
It's crucial that we keep our code as short as possible, even more important than completeness.
Our labeling code is already twice as large as MapBox one, and 5 times larger than Tangram's.

Reply:
True it is a part that I missed after changing implementation to hold LayoutState all the time.